services/config_service.rb

```diff
@@ -28,6 +30,89 @@ def active_code_hosting_service_key(provider_credential: nil)
+
+      run_pending_github_builds(projects: github_projects, github_service: github_service)
+      enqueue_builds_for_open_github_prs_with_no_status(projects: github_projects, github_service: github_service)
+    end
+
+    # In the event of a server crash, we want to run pending builds on server
+    # initialization for all projects for the provider credentials
+    def run_pending_github_builds(projects: nil, github_service: nil)
+      logger.debug("Searching all projects for commits with pending status that need a new build")
+      # For each project, rerun all builds with the status of "pending"
+      projects.each do |project|
+        pending_build_shas_needing_rebuilds = Services.build_service.pending_build_shas_needing_rebuilds(project: project)
+        logger.debug("No pending work to reschedule for #{project.project_name}") if pending_build_shas_needing_rebuilds.count == 0
+
+        # Enqueue each pending build rerun in an asynchronous task queue
+        pending_build_shas_needing_rebuilds.each do |sha|
+          logger.debug("Found sha #{sha} that needs a rebuild for #{project.project_name}")
+          build_runner = FastlaneBuildRunner.new(
+            project: project,
+            sha: sha,
+            github_service: github_service,
+            work_queue: FastlaneCI::GitRepo.git_action_queue # using the git repo queue because of https://github.com/ruby-git/ruby-git/issues/355
+          )
+          build_runner.setup(parameters: nil)
+          Services.build_runner_service.add_build_runner(build_runner: build_runner)
+        end
+      end
+    end
+
+    # We might be in a situation where we have an open pr, but no status yet
+    # if that's the case, we should enqueue a build for it
+    def enqueue_builds_for_open_github_prs_with_no_status(projects: nil, github_service: nil)
+      logger.debug("Searching for open PRs with no status and starting a build for them")
+      projects.each do |project|
+        # TODO: generalize this sort of thing
+        credential_type = project.repo_config.provider_credential_type_needed
+
+        # we don't support anything other than GitHub right now
+        next unless credential_type == FastlaneCI::ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github]
+
+        # Collect all the branches from the triggers on this project that are commit-based
+        branches_to_check = project.job_triggers
+                                   .select { |trigger| trigger.type == FastlaneCI::JobTrigger::TRIGGER_TYPE[:commit] }
+                                   .map(&:branch)
+                                   .uniq
+
+        repo_full_name = project.repo_config.full_name
+        # let's get all commit shas that need a build (no status yet)
+        commit_shas = github_service.last_commit_sha_for_all_open_pull_requests(repo_full_name: repo_full_name, branches: branches_to_check)
```

Comment:
Same here, really, this logic should be broken down between the services and a coordinator. I'm the worst, I know.

Reply:
No, you're not the worst - I'd prefer to not have this coupled so it doesn't cause design headaches in the future.